### ISSUE TITLE
BUG: DataFrame.sort_index() ignores the values returned by the key function

### ISSUE THREAD

**Issue by nick-maykr** [pandas-dev/pandas#56224]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
def sort_function(idx: pd.Index) -> pd.Index:
    print(f"Called with {idx=}")
    return sorted(idx)


df = pd.DataFrame(
    {"col": [1, 9, 4]},
    index=[1, 3, 2]
)
df.sort_index(key=sort_function)
```

### Issue Description

Result:
```
Called with idx=Index([1, 3, 2], dtype='int64')
    col
1	1
3	9
2	4
```

### Expected Behavior

Must be:
```
Called with idx=Index([1, 3, 2], dtype='int64')
    col
1	1
2	4
3	9
```

### Installed Versions

```
>>> pd.__version__
'2.1.3'
```

**Comment by paulreece** [comment]
This isn't a bug.  It's failing because you're sorting it before it has a chance to sort and it exits the sort function early in order to avoid doing unnecessary work.  So it thinks that you have sorted it even though it hasn't.  I'm not sure I see the use case here of altering this function in order to get this to work.

https://github.com/pandas-dev/pandas/blob/ae6d27937b763af21502af839d0a738da6137bb0/pandas/core/sorting.py#L101-L106

**Comment by rhshadrach** [comment]
I think this might be a misunderstanding of what `key` does. Here is an example using just `sorted`:

```
mylist = [(1, 3), (2, 2), (3, 1)]

print(sorted(mylist, key=lambda x: x[0]))
# [(1, 3), (2, 2), (3, 1)]

print(sorted(mylist, key=lambda x: x[1]))
# [(3, 1), (2, 2), (1, 3)]
```

`key` takes the input values and uses them to determine what order to place the results in. You want to think of pandas as doing the same, but vectorized. So it gives you an index `[1, 3, 2]` and you return an index `[1, 2, 3]`. This tells the sort function to put the first element first, the 2nd element 2nd, and the 3rd element 3rd - in other words to do nothing. What you should return is the positions you want the elements to go:

```
def sort_function(idx: pd.Index) -> pd.Index:
    print(f"Called with {idx=}")
    return idx.argsort()


df = pd.DataFrame(
    {"col": [1, 9, 4]},
    index=[1, 3, 2]
)
print(df.sort_index(key=sort_function))
#    col
# 1    1
# 2    4
# 3    9
```

**Comment by nick-maykr** [comment]
@rhshadrach @paulreece You are absolutely right, I misunderstood how the key function works, although the documentation is pretty straightforward: 
key
callable, optional
If not None, apply the key function to the index values **before** sorting.

Sorry for my oversight.

**Comment by paulreece** [comment]
@nick-maykr no worries we are where to help 😎

**Comment by Zalnd** [comment]
> @rhshadrach @paulreece You are absolutely right, I misunderstood how the key function works, although the documentation is pretty straightforward: key callable, optional If not None, apply the key function to the index values **before** sorting.
> 
> Sorry for my oversight.

The real issue here is that the documentation says, "This works similarly to the key argument in Python's built-in sorted() function, with one key difference: the key function must be vectorized. It should take an Index as input and return an Index of the same shape."

When I read that, I had the same thought as you, i.e., `df.sort_index(key=sorted, inplace=True).index` would be equal to `sorted(df.index)`
